sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -517,10 +517,14 @@ struct _pi_context : _pi_object {
   // Mutex Lock for the Command List Cache. This lock is used to control both
   // compute and copy command list caches.
   std::mutex ZeCommandListCacheMutex;
-  // Cache of all currently Available Command Lists for use by PI APIs
-  std::list<ze_command_list_handle_t> ZeComputeCommandListCache;
-  // Cache of all currently Available Copy Command Lists for use by PI APIs
-  std::list<ze_command_list_handle_t> ZeCopyCommandListCache;
+  // Cache of all currently Available Command Lists for use by PI APIs.
+  // Note that command-list can only be re-used on the same device.
+  std::unordered_map<ze_device_handle_t, std::list<ze_command_list_handle_t>>
+      ZeComputeCommandListCache;
+  // Cache of all currently Available Copy Command Lists for use by PI APIs.
+  // Note that command-list can only be re-used on the same device.
+  std::unordered_map<ze_device_handle_t, std::list<ze_command_list_handle_t>>
```

Comment:
are we creating here the same cache for a single root device (so the same cache would be used for its subdevices)?

Reply:
Not really. The "key" to the map is the L0 device on which a SYCL queue was created. It could be a sub-device, in which case other sub-devices wouldn't reuse the command-lists from that cache. I will look into using the "root" device as a key always. Should probably not be hard to do.

Reply:
@jandres742 : I assume the command-list be created on a root-device so it can be used on all of its sub-devices, right? Is there no overhead in creating on a root-device, if eventually it's only used on a specific sub-device?

Reply:
@smaslov-intel  the lists can be executed on any queue of the same device. That's functional, but as you said, there's some overhead.

Imagine a system with multiple subdevices, and you create a list against one subdevice handle, and the queue against another sub-device handle. Functionally speaking, that works. But there could some considerations in terms of performance, if when the list is executed in another sub-device, it needs to access memory in the sub-device where it was originally created (maybe because there was something allocated there at zeCommandListCreate time).

so i think that yes, as you have it right now, having the key as the device handle, and each sub-device with its own list, would provide the best performance.

Reply:
OK. I'll just add a TODO comment about the possibility for the larger re-use at possibly some performance cost.